Feat: Signature order-independence fix for multisig tx

fess-v opened the pull request:
### Description

UPD v2 - removed rolling hash, gating logic implemented for being ready with the Nakamoto release (3.0 epoch)

UPD - check discussion below, decided to leave the rolling hash functionality and remove signatures & public keys from hash creation, add new multisig transaction hash modes following this [SIP](https://github.com/stacksgov/sips/pull/152).
This [comment](https://github.com/stacks-network/stacks-blockchain/pull/3710#issuecomment-1713692399) explains this pull request.

OLD:
After the discussion on the SIP call from May 5th, we decided to make a PR with our ideas for fixing the signing ordering issue for multisig transactions.
Forum post with the initial discussion - https://forum.stacks.org/t/implementing-a-fully-functioning-multisig-wallet-on-stacks/14889
SIP PR - https://github.com/stacksgov/sips/pull/139

This PR implements one possible solution for this problem - removing the rolling hash functionality from signing and verification procedures. This seems to be a redundant security measure that adds a lot of restrictions on the signing flow.
It is still a draft though so return types for verification and signing methods as well as test cases should be revised.

If it's required anyway, there's also another solution that we can think of - to unbind the relation of transaction fields to `StacksAddress::from_public_keys`. It can be done for instance via an additional public keys input array (in the multisig address creation order) or a separate index variable in each signing field.

An example, where the last parameter in the Signature option is the index in the initial address creation public key array:
```rust
pub enum TransactionAuthField {
    PublicKey(StacksPublicKey),
    Signature(TransactionPublicKeyEncoding, MessageSignature, u8),
}
```

In the second scenario, there will still be one restriction on the signing order - other users will have to get the previous signer's signature to continue the signing process. But it won't be critical anymore and will allow to create multisig wallets on Stacks.

### Applicable issues
- https://github.com/stacks-network/stacks-blockchain/issues/2622
- https://github.com/hirosystems/stacks.js/issues/1487

### Additional info (benefits, drawbacks, caveats)
It is a consensus-critical change, therefore should be probably included in the next big release/fork of the network.
Solving this issue will open the multisig functionality to a wide range of use cases.

### Checklist - TODO
- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/stacks-network/stacks-blockchain?pullRequest=3710) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/stacks-network/stacks-blockchain?pullRequest=3710) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/stacks-network/stacks-blockchain?pullRequest=3710) it.</sub>

fess-v commented:
I just added a second draft to the PR using an additional field in the MultisigSpendingCondition struct, it doesn't remove the rolling hash but adds a correct order vector which helps to verify the multisig tx

jcnelson commented:
This is nowhere close to being ready.

First, adding a new mode of transaction authorization requires a hard fork, with gating logic to ensure that any transaction with this new struct will be unparseable until after the hard fork takes place.  It will also require rigorous tests of this gating.  Neither is present in this PR.

Second, the issues linked calls for **order-independent transaction signing and verifying.** As in, if Alice, Bob, and Charlie participate in a 2-of-3 multisig, then the order in which any two or more of them sign ought to result in not only the same multisig address, but also the same signatures (and same intermediate sighashes).  This requirement is not satisfied with this code.

Third, there is zero test coverage for the feature.  Even a basic test such as the one above would have demonstrated that this code does not address the problem at hand.

jcnelson commented:
> Your second point about signatures being equal from different owners is a bit hard to imagine now, could you elaborate on that? Do you mean when Alice signs the transaction as the first owner and when Bob signs as the first owner - they should get the same signature and intermediate sighash as the result? Or you're talking about when Bob signs as the first or the second owner - he should get in both cases the same signature as the outcome? Do you have any ideas on potential implementation logic in any case?

Stacks multisig addresses are calculated the exact same way the are in Bitcoin (see `StacksAddress:;from_public_keys()`).  The order in which the public keys are given matters, just as it does for Bitcoin.  However, unlike Bitcoin, the way signatures are calculated in `MultisigSpendingCondition` forces users to sign the transaction in the same order as their public keys were hashed to produce the multisig address.

Suppose Alice, Bob, and Charlie create a 2-of-3 multisig address, with their public keys given in that order.  Suppose Alice and Charlie sign.  When Alice signs, her signature (but not her public key) is appended to the `MultisigSpendingCondition`'s state.  Alice then adds Bob's public key to the `MultisigSpendingCondition` and gives the partially-signed transaction to Charlie.  Then when Charlie signs, he signs not only the transaction, but also signs Alice's signature. This is because the sighash is calculated from the _current state_ of the `MultisigSpendingCondition` structure -- see `TransactionSpendingCondition::make_sighash_presign()`, `TransactionSpendingCondition::make_sighash_postsign()`, `TransactionSpendingCondition::next_signature()`, and `TransactionSpendingCondition::next_verification()`.

Now, if you look at `MultisigSpendingCondition::verify()`, you'll see that the auth structure would contain Alice's signature, Bob's public key, and Charlie's signature.  Alice's public key is recovered from her signature, Bob's is popped out of the auth structure, and Charlie's public key is recovered from his signature (which means recovering it from the intermediate sighash which commits to Alice's signature).  The three public keys are then hashed to recover a multisig address.  If the multisig address matches the address in the transaction, then the signature is valid.

What needs to happen is that Charlie's signature _must not commit_ to Alice's signature, _but instead_ to just the transaction sighash itself.  While Alice's and Charlie's signatures and Bob's public keys need to be inserted in to the auth structure in the same order in order for `verify()` to keep working, it would now be the case that Charlie no longer needs Alice to sign first.

fess-v commented:
@jcnelson Oh thanks a lot for the explanation!
that's what I tried to do with my approach, but I thought it was rejected 
If you can check the description of the PR and the first commit, I was trying to remove the rolling hash (the thing that you explained in the message) from `TransactionSpendingCondition::make_sighash_presign()`, `TransactionSpendingCondition::make_sighash_postsign()`
Is it the one we need to follow then? Or did I approach it wrong?

fess-v commented:
@jcnelson Yeah I understood that, I don't like this approach either that's why it's the second draft
Could you check the first commit? Is it the way to go? I just deleted the whole logic flow because I thought that this way was declined, but from your explanation, it seems that it's a better way to do that

jcnelson commented:
Hey, I'm sorry for being snippy earlier.  I'm glad you have taken the initiative to implement this feature.  It's not the most straightforward thing to do, either, so kudos to you for taking this on :)

What made me grumpy earlier was all the emails you sent to me and my colleagues to review the PR, despite it being in draft status (i.e. not ready for review).  Please don't do that.  We're all subscribed to the repository already; we already get emails when PRs come in and get commented on.  It sounded like you really, really wanted an assessment despite it not being ready for review.

---

Now to get to your questions:

> Could you check the first commit? Is it the way to go? I just deleted the whole logic flow because I thought that this way was declined, but from your explanation, it seems that it's a better way to do that
> I just thought that the rolling hash thing was required and deleted the first draft

The rolling hash approach is still required, because it provides linear-time signing and verification.  For example, contrast this to Bitcoin's signing and verifying algorithm, which before segwit was quadratic-time.

However, the _contents_ of the rolling hash can be different in order to accommodate new signing protocols.  In particular, you'd make it so that sighash does _not_ contain the previous signers' signatures and public keys.

Insofar as implementation, you'll notice that we have a few different enum variants for spending conditions.  An order-independent multisig spending condition would be its own enum variant, with its own ID byte.  This way, the block validation logic (which is epoch-gated) can be updated to reject blocks with this new spending condition structure before the hard fork in which this feature will ship takes effect.

Within `TransactionSpendingCondition`, you'll need new function bodies for `make_sighash_presign()` and `make_sighash_postsign()` which permit order-independent signatures if the current enum variant calls for order-independent signatures.  On top of this, the implementation will need to implement `StacksMessageCodec` for this variant, and will need encoding/decoding tests.  In addition, you'll need to extend the tests for auth structs to cover this new variant -- both in the positive case (i.e. the signature can be produced and verified) and in the negative case (i.e. corruption to the unsigned data will lead to a signature validation error).

Once you're confident with the new `TransactionSpendingCondition` variant, you'll additionally need to update the block and microblock validation logic so that blocks with this new condition get rejected by the network if they appear prior to the start of the system epoch in which they're allowed.  This will include patching the miner code as well, so the miner does not pick out these transactions from the mempool if they arrive before the hard fork activates.

---

Again, thank you for PR.  If you have any questions or concerns in making progress on the above, please drop a comment here :)

fess-v commented:
Hey @jcnelson! First of all, thanks so much for your comment, now it's clear where to proceed, that's exactly what I was looking for!
Only one question left - what do you think is more appropriate as a base for the new rolling hash? We can, for instance, add to each signature - a signer index from the multi-sig address creation array (i.e. index in the `fields` array from `MultisigSpendingCondition`), but I can't think of anything else as a "rolling" part except signatures and public keys. It will be added in the `make_sighash_presign` method as an additional parameter, but `make_sighash_postsign` won't be used in the new `Spending Condition` since its main purpose is to use the current owner's signature for the next hash. In this case, we won't have to introduce any new input parameters, just what we already have.

And sorry for bothering you with emails, just before that I haven't received any reply within a few months on this PR :(
I know everyone is busy with the Nakamoto upgrade, so I just wanted to get some implementation hints to be ready to proceed.

fess-v commented:
Hey @jcnelson, @jbencin! 
I had to jump on our Stacks multisig release preparations and PRs to Hiro and Stacks.js, and finally back now

Implemented the new logic based on your comments above, supported the recent [SIP](https://github.com/stacksgov/sips/pull/152), added `OrderIndependentMultisigHashMode` based on that, followed the previous repo structure for autotests, gating logic, hash modes, etc.

Now `make_sighash_presign` has an iteration also for public keys, not only for signatures, so now the hash for signatures is defined by index in the initial creation array

For that, I added the `order_independent_sighash_by_index` function to generate hash for signing for any owner by index (still rolling hash but without owner's signature or public key)

`sign_no_append` and append signature methods added for both sponsored and standard transactions - because now it is possible to sign a multisig transaction and then add signatures in the correct order

Gating logic added and tested (both sponsored and standard multisig transactions), used Epoch21 for now as a stub, will change it later to the target activation epoch

I'm planning to add:
1. More tests for wrong order signatures for both new hash modes
2. 3/5 and 3/3 multisig tests (now there's only 2/3 as it was before)
3. More gating logic autotests (not only p2sh, at least 2 more autotests for p2wsh - standard and sponsored)

If you have any questions let me know, looking forward to your feedback!
Thanks

jbencin commented on src/chainstate/stacks/mod.rs:
```diff
@@ -552,6 +584,16 @@ pub struct MultisigSpendingCondition {
 }
 
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct OrderIndependentMultisigSpendingCondition {
```
There is a lot of code duplication involved in implementing this structure and its serialization/deserialization. Consider factoring out the fields shared between the different multisig spending conditions, like this:
```rust
pub struct MultisigTransactionData {
    pub signer: Hash160,
    pub nonce: u64,  // nth authorization from this account
    pub tx_fee: u64, // microSTX/compute rate offered by this account
    pub fields: Vec<TransactionAuthField>,
    pub signatures_required: u16,
}

pub struct MultisigSpendingCondition {
    pub hash_mode: MultisigHashMode,
    pub data: MultisigTransactionData,
}

pub struct OrderIndependentMultisigSpendingCondition {
    pub hash_mode: OrderIndependenMultisigHashMode,
    pub data: MultisigTransactionData,
}
```

jcnelson replied:
I'm going to push back against this.  The existing authentication fields are **consensus-critical**, which means that we should be treating them as read-only code.  It's much safer to implement a new order-independent multisig spending condition from scratch than to try and refactor the existing code.  I'm afraid I'm going to have to insist on this as a condition for merging this PR.  Hope you both understand! :pray:

jbencin commented on src/chainstate/stacks/transaction.rs:
```diff
@@ -1637,7 +1861,7 @@ mod test {
                 bytes: Hash160([0xff; 20]),
             }
-            .into(),
+                .into(),
```
I see a lot of changes like this that are just indentation. Are you using a different `rustfmt` config or did the defaults change?

jbencin reviewed (commented):
Added a couple comments, more to come...

jbencin commented on src/chainstate/stacks/auth.rs:
```diff
@@ -330,6 +332,212 @@ impl MultisigSpendingCondition {
+                        pubkey_encoding,
+                        sigbuf
+                    )?;
+                    cur_sighash = sighash_presign;
```
Updated my SIP proposal to remove repeated computation of `sighash-presign`

jbencin commented:
> The rolling hash approach is still required, because it provides linear-time signing and verification. For example, contrast this to Bitcoin's signing and verifying algorithm, which before segwit was quadratic-time

@jcnelson can you explain how a rolling hash is required for linear-time signing and verification? I would think whether you use a rolling hash or not, if you have *n* participants, you will have to produce and validate *n* signatures, making it an `O(n)` (linear time) process either way

jcnelson commented on stacks-common/src/types/mod.rs:
```diff
@@ -75,6 +75,7 @@ pub enum StacksEpochId {
     Epoch22 = 0x0200f,
     Epoch23 = 0x02014,
     Epoch24 = 0x02019,
+    Epoch25 = 0x0201e,
```
The next epoch is going to be `Epoch30`.  If this PR is ready in time, it will ship with Nakamoto v1 (Stacks 3.0).

jcnelson commented on src/chainstate/stacks/auth.rs:
```diff
@@ -125,9 +128,8 @@ impl StacksMessageCodec for TransactionAuthField {
 }
 
-impl StacksMessageCodec for MultisigSpendingCondition {
+impl StacksMessageCodec for MultisigTransactionData {
```
It's of _paramount importance_ that the existing codec for `MultisigSpendingCondition` behaves the same way before and after this feature activates.  I know it sounds paradoxical, but can you please keep the existing `StacksMessageCodec` in place?  I realize that it will lead to a little bit of code duplication between `MulsitigSpendingCondition` and `MultisigTransactionData`, but I'd rather see that than an accidental consensus bug.  Not changing existing code means not introducing accidental consensus bugs :)

jcnelson commented on src/chainstate/stacks/auth.rs:
```diff
@@ -1913,20 +2238,93 @@ mod test {
-            // key encoding
-            TransactionPublicKeyEncoding::Compressed as u8,
-            // signature
+            0x01,
```
So, I think an older version of `cargo fmt` forced each of these array elements to be on one line.  It's fine with me (encouraged, even) to put all array elements for one serialized field on the same line, or in a 128-ish column block, to improve legibility :)

jcnelson commented on src/chainstate/stacks/db/transactions.rs:
```diff
@@ -8375,6 +8375,7 @@ pub mod test {
                     StacksEpochId::Epoch22 => self.get_stacks_epoch(3),
                     StacksEpochId::Epoch23 => self.get_stacks_epoch(4),
                     StacksEpochId::Epoch24 => self.get_stacks_epoch(5),
+                    StacksEpochId::Epoch25 => self.get_stacks_epoch(6),
```
The `process_transaction()` code needs to be epoch gated as well.  The new order-independent multisig spending condition must not be present in any block prior to epoch 3.0.  That is, only the existing authentication modes are recognized.  Anything else should invalidate the whole block.

fess-v replied:
@jcnelson added logic to this method, used the existing validation method, just added the `quiet` parameter

jcnelson commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -4907,6 +4907,15 @@ impl StacksChainState {
                             receipts.append(&mut clarity_tx.block.initialize_epoch_2_4()?);
                             applied = true;
                         }
+                        StacksEpochId::Epoch25 => {
```
Block-processing and block-decoding needs to be epoch-gated.  A block should not even _decode_ if it has a transaction with the new multisig spending condition.  This is because node without this PR wouldn't be able to decode the block either, and nodes with this patch must behave _exactly_ like the current mainnet nodes up until epoch 3.0 activates.

fess-v replied:
@jcnelson @jbencin one question about this comment
Won't it be enough to add the gating logic to the `process_transaction` and `can_include_tx` methods?
The logic from the first one is used anyway in `process_block_transactions` and `process_microblocks_transactions` so blocks won't pass this check, and the second one is used for mempool transaction validation - added there the gating logic too

Or did you mean adding it also to block deserialization methods? These methods will need to have more dependencies to get the current epoch to conduct this verification, which is probably not appropriate for standard methods like `consensus_deserialize`

jbencin reviewed (commented):
This looks pretty good to me. Definitely make those epoch changes Jude suggested. Also maybe add tests for some different thresholds, like 2-of-9, 5-of-5, or something

fess-v commented:
@jbencin @jcnelson fixed all comments except one, commented on that one too, and added more autotests for different multisig  thresholds

jcnelson commented:
I was speaking more with @kantai about this PR, and a related task for Nakamoto which will see Schnorr signature support for transactions added.  Both changes need the same kinds of gating logic:

* A Stacks block with the new auth structure should not even decode via `consensus_deserialize()` prior to Epoch 3.0
* Same goes for a Stacks transaction 

I think this means that we'll need a trait like the following:

```rust
use stacks_common::codec::Error as CodecError;
use std::io::Read;

pub trait DeserializeWithEpoch {
   fn consensus_deserialize_with_epoch<R>(fd: &mut R, epoch_id: StacksEpochId) -> Result<Self, CodecError>;
}
```

Then, we'd implement `DeserializeWithEpoch` for `StacksTransaction` such that `consensus_deserialize_with_epoch()` does all the same things that the `consensus_deserialize()` does for transactions today, and will return a `CodecError` if the Epoch 3.0 structures are present when `epoch_id < StacksEpochId::Epoch30`.  The `consensus_serialize()` body for `StacksTransaction` would panic -- callers would be obliged to call `consensus_deserialize_with_epoch()` instead of `consensus_deserialize()`.

Similarly, we'd implement `DeserializeWithEpoch` for `StacksBlock`, and move its `consensus_deserialize()` function body to `consensus_deserialize_with_epoch()` and have `consensus_deserialize()` panic instead.

fess-v commented:
> I was speaking more with @kantai about this PR, and a related task for Nakamoto which will see Schnorr signature support for transactions added. Both changes need the same kinds of gating logic:
> 
> * A Stacks block with the new auth structure should not even decode via `consensus_deserialize()` prior to Epoch 3.0
> * Same goes for a Stacks transaction
> 
> I think this means that we'll need a trait like the following:
> 
> ```rust
> use stacks_common::codec::Error as CodecError;
> use std::io::Read;
> 
> pub trait DeserializeWithEpoch {
>    fn consensus_deserialize_with_epoch<R>(fd: &mut R, epoch_id: StacksEpochId) -> Result<Self, CodecError>;
> }
> ```
> 
> Then, we'd implement `DeserializeWithEpoch` for `StacksTransaction` such that `consensus_deserialize_with_epoch()` does all the same things that the `consensus_deserialize()` does for transactions today, and will return a `CodecError` if the Epoch 3.0 structures are present when `epoch_id < StacksEpochId::Epoch30`. The `consensus_serialize()` body for `StacksTransaction` would panic -- callers would be obliged to call `consensus_deserialize_with_epoch()` instead of `consensus_deserialize()`.
> 
> Similarly, we'd implement `DeserializeWithEpoch` for `StacksBlock`, and move its `consensus_deserialize()` function body to `consensus_deserialize_with_epoch()` and have `consensus_deserialize()` panic instead.

Sounds clear, thanks! adding that soon too

fess-v commented:
Hey @jbencin @jcnelson ! I'm currently making a fix regarding the last comment about the gating logic and have a few questions

What's the best way to get the current epoch_id inside https methods like `parse_transaction_unconfirmed` and `parse_postblock`? It is needed for the new `consensus_deserialize_with_epoch` method for both StacksTransaction and StacksBlock

In other scenarios for autotests I'll mostly hardcode it, for cli - add input parameters, for blocks.rs for instance I'll use something similar to:
`let epoch_id = SortitionDB::get_stacks_epoch(sort_tx, burn_header_height)?`

The same question is also for the method `find_next_missing_transactions`, maybe you have something in mind?

Other places seem to have an easy way to get the current epoch id

And one more question - should we do the same thing for `StacksMicroBlock` as well? 
Thanks!

jbencin commented:
> And one more question - should we do the same thing for `StacksMicroBlock` as well?

It is my understanding that microblocks are going away as of 3.0, so I wouldn't worry about it

jbencin commented:
> Then, we'd implement `DeserializeWithEpoch` for `StacksTransaction` such that...

Hey @fess-v I'm going to need this same trait for #3974. Let me know if you've added this already. If not, I'll start work on it.

fess-v commented:
Hey @jbencin I’m nearly done with it, will finish within the next few days

fess-v commented:
Hey @jbencin @jcnelson, while I'm still working on additional autotests, describing the current state:

1. Added DeserializeWithEpoch for both StacksBlock and StacksTransaction 
2. Added #[rustfmt::skip] for the `auth.rs` test module, we discussed above that it should have a custom format for test arrays (fixed in previous commits)
3. Fixed failing autotests 
4. Added DeserializeWithEpoch for StacksMicroblock as well just for autotesting purposes

What's left:
1. Add more autotests for `consensus_deserialize_with_epoch` for both StacksBlock and StacksTransaction for epoch-related logic
2. Check if in all places where `consensus_deserialize_with_epoch` is used, we utilize the correct epoch

For 2nd please take a closer look at:
1. `http.rs` 
2. `mempool.rs`
3. `blockstack_cli.rs` - should we get it from args as we do in `main.rs`?
4. `rpc.rs`

In these files, I just took  `StacksEpochId::latest()` since it's unclear what's the best way to get it there, looking forward to any suggestions

fess-v commented on src/chainstate/stacks/block.rs:
```diff
@@ -375,6 +387,13 @@ impl StacksMessageCodec for StacksBlock {
+            return Err(codec_error::DeserializeError(
+                "Invalid block: target epoch is not activated".to_string(),
+            ));
+        }
```
@jbencin @jcnelson removing this condition now, it is redundant because each tx is checked anyway in `read_next_at_most_with_epoch`

fess-v commented:
@jbencin @jcnelson added autotests, just waiting on feedback regarding my comments above

fess-v commented:
Hey @jcnelson just a reminder about this PR, waiting on any feedback and ready to finalize it

saralab commented:
Waiting for the SIP approval

saralab commented:
Rebase with next

fess-v commented:
@jbencin @jcnelson 
Just added more autotests, what's left:
1. main.rs & blockstack_cli.rs -> discuss if we need to add an epoch parameter (optional or required), added as an example to main.rs as an optional one.
2. API methods (described above) -> add new API methods that will accept an epoch as a parameter (where it's needed, for `StacksBlock` and `StacksTransaction`), old methods need to restrict so they use Epoch25 with no breaking logic for previous epochs. This point should be discussed, more details can be found in the message above.
3. Documentation fixes. Will work on it once we finalize the logic flow.

jbencin commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -831,7 +832,7 @@ impl NakamotoBlock {
             }
         }
-        if !StacksBlock::validate_transactions_static_epoch(&self.txs, epoch_id) {
+        if !StacksBlock::validate_transactions_static_epoch(&self.txs, epoch_id, true) {
```
Why are we muting errors here?

jbencin commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -2714,12 +2723,22 @@ impl StacksMessageCodec for NakamotoBlock {
 
     fn consensus_deserialize<R: std::io::Read>(fd: &mut R) -> Result<Self, CodecError> {
-        let (header, txs) = {
+        panic!("NakamotoBlock should be deserialized with consensus_deserialize_with_epoch instead")
```
I understand this function exists here because you need to `impl StacksMessageCodec` to get `consensus_serialize()`, but maybe we can separate `consensus_serialize()` and `consensus_deserialize()` into their own traits, and then have:

```rust
pub trait StacksMessageCodec: Serialize + Deserialize {}
pub trait StacksMessageCodecWithEpoch: Serialize + DeserializeWithEpoch {}
```

There will be other structs implementing `DeserializeWithEpoch`, so it's worth having to not have a bunch of potential panics in the code.

This will require replacing a lot of references to `StacksMessageCodec` with the more specific traits. Maybe this is a task for future PR...

jcnelson replied:
I'd drop this change.  You don't need `DeserializeWithEpoch` for `NakamotoBlock` since `NakamotoBlock` supports all transaction types anyway.

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,210 @@ impl MultisigSpendingCondition {
+        }
+
+        // must be given the right number of signatures
+        if num_sigs_given != signatures_required {
```
I think we should accept a transaction if it has more then the required number of signers:

```suggestion
        if num_sigs_given < signatures_required {
```

I can think of a scenario where this might happen: Let's say some org has a 3-of-5 multisig account, meaning they can spend funds with a majority vote. Maybe they want to submit a transaction with 4-of-5 or 5-of-5 signatures, even though it's not necessary to spend the funds, in order to leave a record of the vote on the blockchain

fess-v replied:
This is logical and I agree with the point
It will be different from sequential multisig tx though (since it has != while this will have <), so we need to make a decision here, same thing goes for the comment below @jcnelson @jbencin

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,210 @@ impl MultisigSpendingCondition {
+        Ok(())
+    }
+
+    fn consensus_deserialize<R: Read>(
```
Shouldn't this implement `DeserializeWithEpoch` and panic on `consensus_deserialize()`?

jcnelson replied:
It's fine without it as long as the block validation checks catch it.

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,210 @@ impl MultisigSpendingCondition {
+        cond_code: &TransactionAuthFlags,
+    ) -> Result<Txid, net_error> {
+        let mut pubkeys = vec![];
+        let cur_sighash = initial_sighash.clone();
```
Let's get rid of `cur_sighash` since it never changes now and just use `initial_sighash`

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,210 @@ impl MultisigSpendingCondition {
+            pubkeys.push(pubkey);
+        }
+
+        if num_sigs != self.signatures_required {
```
Again, should we allow for more signers than the minimum threshold?

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -908,9 +1205,35 @@ impl TransactionAuth {
+        privks: &[StacksPrivateKey],
+        num_sigs: u16,
+    ) -> Option<TransactionAuth> {
+        let mut pubks = vec![];
+        for privk in privks.iter() {
+            pubks.push(StacksPublicKey::from_private(privk));
+        }
```
Probably more efficient to do

```suggestion
        let pubks = privks.iter()
            .map(|privk| StacksPublicKey::from_private(privk))
            .collect();
```

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -908,9 +1205,35 @@ impl TransactionAuth {
+        privks: &[StacksPrivateKey],
+        num_sigs: u16,
+    ) -> Option<TransactionAuth> {
+        let mut pubks = vec![];
+        for privk in privks.iter() {
+            pubks.push(StacksPublicKey::from_private(privk));
+        }
```
Same as above:

```suggestion
         let pubks = privks.iter()
            .map(|privk| StacksPublicKey::from_private(privk))
            .collect();
```

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -1854,139 +1590,153 @@ mod test {
+            // filed #2: signature
+            0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xfe, 0xf ...
+            // field #3: public key
+            TransactionAuthFieldID::PublicKeyUncompressed as u8,
+            // field #3: key (compressed)
```
Is this compressed or not?

fess-v replied:
fixed

jbencin commented on stackslib/src/chainstate/stacks/mod.rs:
```diff
@@ -544,6 +551,35 @@ impl MultisigHashMode {
+        }
+    }
+
+    pub fn from_u8(n: u8) -> Option<OrderIndependentMultisigHashMode> {
```
I know you're just following patterns here, but it's better to implement `TryFrom` for conversions that can fail, than make your own function and return an `Option`

jbencin commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -567,38 +577,86 @@ impl StacksBlock {
                 if proof_opt.is_some() && epoch_id < StacksEpochId::Epoch30 {
                     // not supported
-                    error!("Coinbase with VRF proof not supported before Stacks 3.0"; "txid" => %tx.txid());
+                    if !quiet {
```
Maybe use a macro to avoid all these `if` statements:

```rust
// Quietable error
macro_rules! qerror {
    ($cond: expr, $($arg:tt)*) => ({
        if ($cond) {
            error!($($arg)*);
        }
    });
}
```

```rust
qerror!(quiet, "Coinbase with VRF proof is required in Stacks 3.0 and later"; "txid" => %tx.txid());
```

jbencin commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -1672,6 +1801,146 @@ mod test {
 
+    fn verify_block_epoch_validation(
+        txs: &[StacksTransaction],
+        tx_coinbase_old: StacksTransaction,
+        tx_coinbase_nakamoto: StacksTransaction,
+        activation_epoch_id: StacksEpochId,
+        header: StacksBlockHeader,
+        need_to_include_coinbase_old: bool,
+        need_to_include_coinbase_nakamoto: bool,
```
Why not this?
```suggestion
        tx_coinbase_old: Option<StacksTransaction>,
        tx_coinbase_nakamoto: Option<StacksTransaction>,
        activation_epoch_id: StacksEpochId,
        header: StacksBlockHeader,
```

jcnelson replied:
Agreed; seconded

jbencin commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -1699,6 +1973,135 @@ mod test {
+            ),
+        );
+
+        let mut tx_signer =
```
It looks like most if not all of our multisig tests are testing using 2-of-3. How about a little variety? 2-of-2, 1-of-4, etc?

jbencin replied:
Never mind, there's a ton of those in `transaction.rs`

jbencin commented on stackslib/src/main.rs:
```diff
@@ -215,13 +216,17 @@ fn main() {
 
         let block_path = &argv[2];
         let block_data = fs::read(block_path).expect(&format!("Failed to open {}", block_path));
+        let epoch_id = parse_input_epoch(3);
```
It seems like none of these commands or their args are documented in `stacks-node --help`. This should be fixed in a separate PR

jbencin commented on stackslib/src/main.rs:
```diff
@@ -1246,6 +1255,7 @@ simulating a miner.
     let events_file = &argv[3];
     let mine_tip_height: u64 = argv[4].parse().expect("Could not parse mine_tip_height");
     let mine_max_txns: u64 = argv[5].parse().expect("Could not parse mine-num-txns");
+    let epoch_id = parse_input_epoch(6);
```
In this subcommand we should probably take the `epoch_id` from the database

jbencin commented on stackslib/src/main.rs:
```diff
@@ -1459,3 +1471,23 @@ simulating a miner.
+            .parse::<u32>()
+            .expect("Failed to parse epoch id");
+        epoch_id = StacksEpochId::try_from(epoch_id_u32)
+            .map_err(|_e| {
+                eprintln!("Failed to match epoch number");
+                process::exit(1);
+            })
+            .unwrap();
```
`exit()` then `unwrap()`? Why not just `expect()`?

jbencin commented on stackslib/src/chainstate/stacks/db/transactions.rs:
```diff
@@ -9348,6 +9359,28 @@ pub mod test {
         let mut chainstate =
             instantiate_chainstate_with_balances(false, 0x80000000, function_name!(), balances);
 
+        let mut tx_runtime_checkerror_trait_no_version = StacksTransaction::new(
```
Can you explain the changes in this file and how they relate to order independent multisig?

fess-v replied:
Because this method uses `process_transaction` method, which is now gated, it is important to fix autotests to match the previous behavior and separate them by `version_opt` parameter, which is also gated

jbencin commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -707,19 +710,24 @@ impl StacksTransaction {
-            },
-            fd.num_read(),
-        ))
+        if !StacksBlock::validate_transactions_static_epoch(&vec![tx.clone()], epoch_id, false) {
```
It's not necessary to use a `Vec` here, use a slice and avoid the heap allocation:

```suggestion
        if !StacksBlock::validate_transactions_static_epoch(&[tx.clone()], epoch_id, false) {
```

This applies to other uses of this function as well

jbencin commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -1350,6 +1522,21 @@ mod test {
+            TransactionSpendingCondition::OrderIndependentMultisig(ref data) => {
+                let mut j = 0;
+                for f in 0..data.fields.len() {
+                    match data.fields[f] {
+                        TransactionAuthField::Signature(_, _) => {
+                            j = f;
+                            break;
+                        }
+                        _ => {
+                            continue;
+                        }
+                    }
```
```suggestion
                    if matches!(data.fields[f], TransactionAuthField::Signature(..)) {
                        j = f;
                        break;
                    }
```

jbencin commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -1371,6 +1558,21 @@ mod test {
+            TransactionSpendingCondition::OrderIndependentMultisig(ref data) => {
+                let mut j = 0;
+                for f in 0..data.fields.len() {
+                    match data.fields[f] {
+                        TransactionAuthField::PublicKey(_) => {
+                            j = f;
+                            break;
+                        }
+                        _ => {
+                            continue;
+                        }
+                    }
```
```suggestion
                    if matches!(data.fields[f], TransactionAuthField::PublicKey(..)) {
                        j = f;
                        break;
                    }
```

jbencin commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -951,6 +986,98 @@ impl StacksTransaction {
         Ok(next_sighash)
     }
 
+    pub fn sign_no_append_origin(
```
It looks like this and the next 3 functions are only used in test code. Can you add `#[cfg(test)]` to them?

jbencin commented on testnet/stacks-node/src/tests/mempool.rs:
```diff
@@ -237,8 +237,11 @@ fn mempool_setup_chainstate() {
-                    StacksTransaction::consensus_deserialize(&mut tx_bytes.as_slice()).unwrap();
+                let tx = StacksTransaction::consensus_deserialize_with_epoch(
+                    &mut tx_bytes.as_slice(),
+                    StacksEpochId::Epoch25,
```
Should we be using `StacksEpochId::latest()` in this file?

jbencin commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -804,7 +862,66 @@ impl StacksMessageCodec for StacksMicroblock {
             let mut bound_read = BoundReader::from_reader(fd, MAX_MESSAGE_LEN as u64);
-            read_next(&mut bound_read)
+            // The latest epoch where StacksMicroblock exist is Epoch25
+            read_next_at_most_with_epoch(&mut bound_read, u32::MAX, StacksEpochId::Epoch25)
```
Should this use the `epoch_id` of the current scope?
```suggestion
            read_next_at_most_with_epoch(&mut bound_read, u32::MAX, epoch_id)
```

jcnelson replied:
I actually don't think this call is necessary at all.  You are already going to do epoch-specific checks in this function to determine if the transactions are _semantically_ valid within the epoch, so you should just decode them with `read_next_at_most()` and then deal with them below.

jcnelson commented on libsigner/src/messages.rs:
```diff
@@ -362,7 +363,11 @@ impl StacksMessageCodec for SignerMessage {
             }
             SignerMessageTypePrefix::Transactions => {
-                let transactions = read_next::<Vec<StacksTransaction>, _>(fd)?;
+                // I don't think these messages are stored on the blockchain, so `StacksEpochId::latest()` should be fine
```
Yes the are (or rather will be). Please use the current system epoch.

jcnelson replied:
Specifically, you should filter out transactions that would not be mineable in the current epoch, and keep the rest.  These are transactions that the signers can require the miner to include in their blocks, so the signer should not be able to compel the miner to mine invalid transactions.

jcnelson commented on stacks-common/src/codec/mod.rs:
```diff
@@ -183,6 +193,66 @@ pub fn read_next_exact<R: Read, T: StacksMessageCodec + Sized>(
     read_next_vec::<T, R>(fd, num_items, 0)
 }
 
+pub fn read_next_with_epoch<T: DeserializeWithEpoch, R: Read>(
```
I don't think this method or `read_next_vec_with_epoch` are necessary or desirable.  Codec implementations that need to be epoch-aware should directly inspect the data they decode and decide how to proceed from there.

jcnelson commented on stackslib/src/blockstack_cli.rs:
```diff
@@ -314,8 +315,10 @@ fn sign_transaction_single_sig_standard(
 ) -> Result<StacksTransaction, CliError> {
-    let transaction =
-        StacksTransaction::consensus_deserialize(&mut io::Cursor::new(&hex_bytes(transaction)?))?;
+    let transaction = StacksTransaction::consensus_deserialize_with_epoch(
```
If you're going to use `StacksEpochId::latest()`, then you can just use `consensus_deserialize`

jcnelson commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -3357,12 +3357,22 @@ impl StacksMessageCodec for NakamotoBlock {
+    }
+}
+
+impl DeserializeWithEpoch for NakamotoBlock {
```
Why is this trait implementation necessary?

jcnelson commented on stackslib/src/chainstate/nakamoto/staging_blocks.rs:
```diff
@@ -254,7 +258,10 @@ impl<'a> NakamotoStagingBlocksConnRef<'a> {
             return Ok(None);
         };
-        let block = NakamotoBlock::consensus_deserialize(&mut block_bytes.as_slice())?;
+        let block = NakamotoBlock::consensus_deserialize_with_epoch(
```
`consensus_deserialize` is fine

jcnelson commented on stackslib/src/chainstate/nakamoto/staging_blocks.rs:
```diff
@@ -304,7 +311,7 @@ impl<'a> NakamotoStagingBlocksConnRef<'a> {
             .query_row_and_then(query, NO_PARAMS, |row| {
                 let data: Vec<u8> = row.get("data")?;
-                let block = NakamotoBlock::consensus_deserialize(&mut data.as_slice())?;
+                let block = NakamotoBlock::consensus_deserialize_with_epoch(&mut data.as_slice(), StacksEpochId::latest())?;
```
`consensus_deserialize` is fine

jcnelson commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -40,6 +41,15 @@ use crate::chainstate::stacks::{Error, StacksBlockHeader, StacksMicroblockHeader
 use crate::net::Error as net_error;
 
+/// Quietable error
+macro_rules! qerror {
```
Please drop this macro altogether.  If something is `error!`-level, it should never be quieted.

jcnelson commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -804,7 +858,66 @@ impl StacksMessageCodec for StacksMicroblock {
+}
+
+// This implementation is used for testing purposes, StacksMicroblock won't be used in Epoch 3.0
+impl DeserializeWithEpoch for StacksMicroblock {
```
I don't think you need this.  StacksMicroblocks are disabled in epoch 2.5

jcnelson commented on stackslib/src/chainstate/stacks/db/blocks.rs:
```diff
@@ -560,6 +560,28 @@ impl StacksChainState {
         Ok(inst)
     }
 
+    pub fn consensus_load_with_epoch<T: DeserializeWithEpoch>(
```
Stacks blocks are only stored to disk up through epoch 2.5, so this is not needed

jcnelson commented on stackslib/src/chainstate/stacks/db/blocks.rs:
```diff
@@ -827,6 +849,20 @@ impl StacksChainState {
+        )
+    }
+
+    pub fn load_block_with_epoch(
```
Also not needed

jcnelson commented on stackslib/src/chainstate/stacks/db/transactions.rs:
```diff
@@ -1479,6 +1479,17 @@ impl StacksChainState {
                 return Err(Error::InvalidStacksTransaction(msg, false));
             }
         }
+        let is_transaction_valid_in_epoch =
```
Nit: drop this let-binding and just put this call into the `if` statement

jcnelson commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -686,19 +689,24 @@ impl StacksTransaction {
-            fd.num_read(),
-        ))
+        if !StacksBlock::validate_transactions_static_epoch(&[tx.clone()], epoch_id, false) {
+            warn!("Invalid tx: target epoch is not activated");
```
I'd drop this entirely.  It should be fine to decode a transaction regardless of what epoch we're in.  What ultimately matters is that we don't process a transaction that isn't supported in the current epoch.  To support that, we would also treat a block as invalid if it contained an unsupported transaction (so we don't even store it or relay it).

jcnelson commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -734,8 +742,17 @@ impl StacksMessageCodec for StacksTransaction {
+    }
+}
+
+impl DeserializeWithEpoch for StacksTransaction {
```
You don't need this

jcnelson commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -938,6 +973,101 @@ impl StacksTransaction {
     }
 
+    #[cfg(any(test, feature = "testing"))]
+    fn sign_no_append_origin(
```
Test code should go into the test module

jcnelson commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -938,6 +973,101 @@ impl StacksTransaction {
+        Ok(next_sig)
+    }
+
+    pub fn append_sponsor_signature(
```
Missing docstring

jcnelson commented on stackslib/src/core/mempool.rs:
```diff
@@ -2449,8 +2461,11 @@ impl MemPoolDB {
     ) -> Result<(), MemPoolRejection> {
-        let tx = StacksTransaction::consensus_deserialize(&mut &tx_bytes[..])
-            .map_err(MemPoolRejection::DeserializationFailure)?;
+        let tx = StacksTransaction::consensus_deserialize_with_epoch(
```
Broadly speaking, you don't want this method to be used anywhere in the mempool

jcnelson commented on stackslib/src/main.rs:
```diff
@@ -195,8 +195,9 @@ fn main() {
+        let epoch_id = parse_input_epoch(3);
 
-        let tx = StacksTransaction::consensus_deserialize(&mut debug_cursor)
+        let tx = StacksTransaction::consensus_deserialize_with_epoch(&mut debug_cursor, epoch_id)
```
Please revert these changes. This tool is used to decode any block given to it.

jcnelson commented on stackslib/src/main.rs:
```diff
@@ -1645,7 +1681,12 @@ fn replay_block(stacks_path: &str, index_block_hash_hex: &str) {
     };
 
-    let block = StacksChainState::extract_stacks_block(&next_staging_block).unwrap();
+    let epoch_id = SortitionDB::get_stacks_epoch(&sort_tx, u64::from(burn_header_height))
```
No need for this -- the block has already been processed so there's no need to revalidate it with the epoch

jcnelson commented on stackslib/src/net/api/getblock.rs:
```diff
@@ -302,7 +303,22 @@ impl StacksHttpResponse {
         // contents will be raw bytes
         let block_bytes: Vec<u8> = contents.try_into()?;
-        let block = StacksBlock::consensus_deserialize(&mut &block_bytes[..])?;
+        let block = StacksBlock::consensus_deserialize_with_epoch(
```
Revert -- StacksBlocks are only accepted in epoch 2.5 or earlier as it is

jcnelson commented on stackslib/src/net/api/gettenure.rs:
```diff
@@ -355,7 +356,8 @@ impl StacksHttpResponse {
         let mut blocks = vec![];
         while ptr.len() > 0 {
-            let block = NakamotoBlock::consensus_deserialize(ptr)?;
+            let block =
```
Revert -- Nakamoto blocks are only considered in epoch 3.0 anyway

jcnelson commented on stackslib/src/net/api/postblock.rs:
```diff
@@ -74,13 +74,18 @@ impl RPCPostBlockRequestHandler {
-            }
-        })?;
+        let block =
+            StacksBlock::consensus_deserialize_with_epoch(&mut body, StacksEpochId::Epoch25)
```
Revert -- Stacks blocks are only valid in epoch 2.5 or earlier

jcnelson commented on stackslib/src/net/api/posttransaction.rs:
```diff
@@ -68,13 +70,18 @@ impl RPCPostTransactionRequestHandler {
-            }
-        })?;
+        let tx =
+            StacksTransaction::consensus_deserialize_with_epoch(&mut body, StacksEpochId::latest())
```
Revert -- we want to accept all syntactically valid transactions into the mempool since they may become semantically valid at an upcoming epoch boundary

jcnelson commented on stackslib/src/net/api/posttransaction.rs:
```diff
@@ -88,7 +95,11 @@ impl RPCPostTransactionRequestHandler {
             let tx_bytes = hex_bytes(&body.tx)
                 .map_err(|_e| Error::DecodeError("Failed to parse tx".into()))?;
-            StacksTransaction::consensus_deserialize(&mut &tx_bytes[..]).map_err(|e| {
+            StacksTransaction::consensus_deserialize_with_epoch(
```
Revert

jcnelson commented on stackslib/src/net/api/tests/getblock.rs:
```diff
@@ -159,7 +161,11 @@ fn test_stream_blocks() {
 
     // should decode back into the block
-    let staging_block = StacksBlock::consensus_deserialize(&mut &all_block_bytes[..]).unwrap();
+    let staging_block = StacksBlock::consensus_deserialize_with_epoch(
```
Revert

jcnelson commented on stackslib/src/net/api/tests/postmempoolquery.rs:
```diff
@@ -214,23 +214,24 @@ fn test_stream_mempool_txs() {
-                        eprintln!("out of transactions");
-                        break;
-                    }
+        let tx: StacksTransaction =
```
Revert

jcnelson commented on stackslib/src/net/relay.rs:
```diff
@@ -5823,21 +5827,8 @@ pub mod test {
 
             let sortdb = peer.sortdb.take().unwrap();
             let mut node = peer.stacks_node.take().unwrap();
-            match Relayer::process_new_anchored_block(
```
Why is this dropped?

jcnelson commented on stackslib/src/net/relay.rs:
```diff
@@ -6003,22 +5996,8 @@ pub mod test {
 
             let sortdb = peer.sortdb.take().unwrap();
             let mut node = peer.stacks_node.take().unwrap();
-            match Relayer::process_new_anchored_block(
```
Why is this dropped?

fess-v replied:
won't be dropped anymore since we removed the epoch gating logic

fess-v replied:
Actually I'm wrong on this one
`make_anchored_block_and_microblock_from_txs` this function filters the incorrect transaction (skips it) because it has new gating logic for transaction processing, therefore the final block won't be rejected
That's why I removed the logic for block processing and checked the amount of transactions left instead, but let me know if you have better ideas on how we can modify this part

jcnelson replied:
In that case, this must be restored.  The code here is checking that block-processing catches blocks with invalid Stacks transactions.  This snippet of test code performs that check.

fess-v replied:
@jcnelson the incorrect transaction is filtered in the step before, `make_anchored_block_and_microblock_from_txs ` this function is being called in the `make_anchored_block_from_txs` function
Which has inside `match builder.try_mine_tx` which rejects the transaction and does not include it in the end because of the gating logic (Error::InvalidStacksTransaction)
Therefore the code I removed won't have an error inside, while I added a line that checks if the incorrect tx was filtered
Do you think we should somehow modify the `make_anchored_block_and_microblock_from_txs ` function or make a new one for this particular autotest? Would be great to get more info on that, thanks

jcnelson replied:
As I said before, the purpose of this dropped code is to verify that block-processing catches blocks with invalid Stacks transactions.  If your changes make it so that it's no longer possible to construct Stacks blocks with invalid transactions such that this test continues to pass as-is, then the correct thing to do is to _fix your patch_ instead of delete my tests. 
 The function `try_mine_tx()` should not be performing an epoch gate check.

The way the system has historically prevented transactions from an as-of-yet-unsupported epoch from finding their way into blocks is to prevent them from entering the mempool.  I see you have already added the gating logic to the function `can_include_tx()`, which is sufficient.

jcnelson commented on testnet/stacks-node/src/tests/epoch_21.rs:
```diff
@@ -384,9 +385,10 @@ fn transition_adds_burn_block_height() {
-    let cc_txid = StacksTransaction::consensus_deserialize(&mut &tx[..])
-        .unwrap()
-        .txid();
+    let cc_txid =
```
Revert all of these changes please

jcnelson commented on testnet/stacks-node/src/tests/epoch_22.rs:
```diff
@@ -540,8 +541,11 @@ fn disable_pox() {
             let tx_bytes = hex_bytes(&raw_tx[2..]).unwrap();
-            let parsed =
-                StacksTransaction::consensus_deserialize(&mut tx_bytes.as_slice()).unwrap();
+            let parsed = StacksTransaction::consensus_deserialize_with_epoch(
```
Revert all of these changes please

jcnelson commented on testnet/stacks-node/src/tests/epoch_23.rs:
```diff
@@ -508,8 +509,11 @@ fn trait_invocation_behavior() {
             let tx_bytes = hex_bytes(&raw_tx[2..]).unwrap();
-            let parsed =
-                StacksTransaction::consensus_deserialize(&mut tx_bytes.as_slice()).unwrap();
+            let parsed = StacksTransaction::consensus_deserialize_with_epoch(
```
Revert

jcnelson commented on testnet/stacks-node/src/tests/epoch_24.rs:
```diff
@@ -645,8 +645,11 @@ fn fix_to_pox_contract() {
             let tx_bytes = hex_bytes(&raw_tx[2..]).unwrap();
-            let parsed =
-                StacksTransaction::consensus_deserialize(&mut tx_bytes.as_slice()).unwrap();
+            let parsed = StacksTransaction::consensus_deserialize_with_epoch(
```
Revert

jcnelson commented on testnet/stacks-node/src/tests/integrations.rs:
```diff
@@ -808,7 +808,7 @@ fn integration_test_get_info() {
                     .unwrap();
 
-                assert_eq!(res, format!("{}", StacksTransaction::consensus_deserialize(&mut &tx_xfer[..]).unwrap().txid()));
+                assert_eq!(res, format!("{}", StacksTransaction::consensus_deserialize_with_epoch(&mut &tx_xfer[..], StacksEpochId::latest()).unwrap().txid()));
```
Revert al of these changes please

jcnelson commented on testnet/stacks-node/src/tests/mempool.rs:
```diff
@@ -236,8 +236,11 @@ fn mempool_setup_chainstate() {
                     make_contract_publish(&contract_sk, 5, 1000, "bar_contract", FOO_CONTRACT);
-                let tx =
-                    StacksTransaction::consensus_deserialize(&mut tx_bytes.as_slice()).unwrap();
+                let tx = StacksTransaction::consensus_deserialize_with_epoch(
```
Revert all of these changes please

jcnelson commented on testnet/stacks-node/src/tests/mod.rs:
```diff
@@ -464,14 +464,17 @@ fn make_microblock(
             let raw_tx = tx.get("raw_tx").unwrap().as_str().unwrap();
             let tx_bytes = hex_bytes(&raw_tx[2..]).unwrap();
-            let parsed = StacksTransaction::consensus_deserialize(&mut &tx_bytes[..]).unwrap();
+            let parsed =
```
This change shouldn't be necessary since these blocks are already mined and processed

jcnelson commented on testnet/stacks-node/src/tests/neon_integrations.rs:
```diff
@@ -721,10 +721,13 @@ pub fn submit_tx(http_origin: &str, tx: &Vec<u8>) -> String {
-            StacksTransaction::consensus_deserialize(&mut &tx[..])
-                .unwrap()
-                .txid()
-                .to_string()
```
Revert all of these changes please

jcnelson reviewed (commented):
I just did a pass over just the epoch-gating, and I think almost all of it can be reverted.  This is a good thing, since it would substantially reduce the changeset and thus reduce the opportunity for consensus bugs to slip in.

After looking through how the implementation adds the new transaction authorization structures, it would appear that the only place where it's necessary to do epoch-gating is in the `StacksBlock::validate_transactions_static_epoch()` function.  This function should be modified to determine whether or not the transaction uses the new transaction authorization, and if so, outright reject the block (since it'll never be supported in epoch 2.5 or earlier, meaning that there will never be a `StacksBlock` that contains such a transaction).  All other epoch-gating logic, including trait implementations for `DeserializeWithEpoch` and all code paths that depend on this trait's existence, can be dropped.

jbencin commented on stackslib/src/net/api/getblock.rs:
```diff
@@ -23,6 +23,7 @@ use serde::de::Error as de_Error;
 use stacks_common::codec::{StacksMessageCodec, MAX_MESSAGE_LEN};
 use stacks_common::types::chainstate::StacksBlockId;
 use stacks_common::types::net::PeerHost;
+use stacks_common::types::StacksEpochId;
```
```suggestion
```

jbencin commented on stackslib/src/net/api/getblock_v3.rs:
```diff
@@ -23,6 +23,7 @@ use serde::de::Error as de_Error;
 use stacks_common::codec::{StacksMessageCodec, MAX_MESSAGE_LEN};
 use stacks_common::types::chainstate::{ConsensusHash, StacksBlockId};
 use stacks_common::types::net::PeerHost;
+use stacks_common::types::StacksEpochId;
```
```suggestion
```

jbencin commented on stackslib/src/net/api/gettenure.rs:
```diff
@@ -22,6 +22,7 @@ use serde::de::Error as de_Error;
 use stacks_common::codec::{StacksMessageCodec, MAX_MESSAGE_LEN};
 use stacks_common::types::chainstate::{ConsensusHash, StacksBlockId};
 use stacks_common::types::net::PeerHost;
+use stacks_common::types::StacksEpochId;
```
```suggestion
```

jbencin commented on stackslib/src/net/api/postblock.rs:
```diff
@@ -23,7 +23,7 @@ use stacks_common::types::chainstate::{
 };
 use stacks_common::types::net::PeerHost;
-use stacks_common::types::StacksPublicKeyBuffer;
+use stacks_common::types::{StacksEpochId, StacksPublicKeyBuffer};
```
```suggestion
use stacks_common::types::StacksPublicKeyBuffer;
```

jbencin commented on stackslib/src/net/api/posttransaction.rs:
```diff
@@ -23,7 +23,7 @@ use stacks_common::types::chainstate::{
 };
 use stacks_common::types::net::PeerHost;
-use stacks_common::types::StacksPublicKeyBuffer;
+use stacks_common::types::{StacksEpochId, StacksPublicKeyBuffer};
```
```suggestion
use stacks_common::types::StacksPublicKeyBuffer;
```

jbencin commented on stackslib/src/net/api/tests/getblock.rs:
```diff
@@ -18,12 +18,11 @@ use std::net::{IpAddr, Ipv4Addr, SocketAddr};
 };
 use stacks_common::types::net::PeerHost;
-use stacks_common::types::Address;
+use stacks_common::types::{Address, StacksEpochId};
```
```suggestion
use stacks_common::types::Address;
```

jbencin commented on stackslib/src/net/api/tests/getblock_v3.rs:
```diff
@@ -23,7 +23,7 @@ use stacks_common::types::chainstate::{
 };
 use stacks_common::types::net::PeerHost;
-use stacks_common::types::Address;
+use stacks_common::types::{Address, StacksEpochId};
```
```suggestion
use stacks_common::types::Address;
```

jbencin commented on stackslib/src/net/api/tests/gettenure.rs:
```diff
@@ -23,7 +23,7 @@ use stacks_common::types::chainstate::{
 };
 use stacks_common::types::net::PeerHost;
-use stacks_common::types::Address;
+use stacks_common::types::{Address, StacksEpochId};
```
```suggestion
use stacks_common::types::Address;
```

jbencin commented on stackslib/src/net/codec.rs:
```diff
@@ -1555,6 +1555,7 @@ impl ProtocolFamily for StacksP2P {
 pub mod test {
     use stacks_common::bitvec::BitVec;
     use stacks_common::codec::NEIGHBOR_ADDRESS_ENCODED_SIZE;
+    use stacks_common::types::StacksEpochId;
```
```suggestion
```

jbencin reviewed (approved):
There's a few things that are not needed after removing `DeserializeWithEpoch`

Otherwise this LGTM and I'd love to get it merged ASAP

jcnelson commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -3357,12 +3357,13 @@ impl StacksMessageCodec for NakamotoBlock {
-            let txs: Vec<_> = read_next(&mut bound_read)?;
-            (header, txs)
-        };
+            // The latest epoch where StacksMicroblock exist is Epoch25
```
What do microblocks have to do with Nakamoto blocks here?

jcnelson commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -3357,12 +3357,13 @@ impl StacksMessageCodec for NakamotoBlock {
-            (header, txs)
-        };
+            // The latest epoch where StacksMicroblock exist is Epoch25
+            read_next_at_most(&mut bound_read, u32::MAX)
```
You can use `read_next()` here because `BoundReader` will stop you from reading more than `MAX_MESSAGE_LEN` bytes

jcnelson commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,204 @@ impl MultisigSpendingCondition {
+            let msg = format!(
+                "Failed to deserialize order independent multisig spending condition: expected compressed keys only"
+            );
+            test_debug!("{msg}");
```
Perhaps this should be a `warn!`?

jbencin replied:
It's `debug` level in `MultisigSpendingCondition`. Actually, all the logs in this file are `debug` level

jcnelson commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,204 @@ impl MultisigSpendingCondition {
+            let msg = format!(
+                "Failed to deserialize order independent multisig spending condition: got {num_sigs_given} sigs, expected at least {signatures_required}"
+            );
+            test_debug!("{msg}");
```
Perhaps this should be a `warn!`?

jbencin replied:
It's `debug` level in `MultisigSpendingCondition`. Actually, all the logs in this file are `debug` level

jcnelson replied:
It probably should be `warn!` there as well

jcnelson replied:
Let's leave it debug for now.  I was hoping we could make it `warn!` because such a problem would invalidate the block, but it'd also end up being very chatty (and log-exhaustive) if someone decided to repeatedly post bad transactions to /v2/transactions.

jcnelson commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -1102,112 +1414,15 @@ mod test {
-            0x11,
-            0x11,
-            0x11,
+            0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11,
```
Thanks for cleaning up this unhelpful artifact of `rustfmt`.  Makes the test a lot more legible :+1:

jcnelson commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,204 @@ impl MultisigSpendingCondition {
+            pubkeys.push(pubkey);
+        }
+
+        if num_sigs < self.signatures_required {
```
While I like the idea of supporting more than the required number of signatures, I'm not sure this code is handling the case where there are a surplus number of signatures.  In particular, if there are more than the minimum number of signatures, then at least one signature would need to decode to a public key that's already present in `fields`, right?  If so, then this check appears to be absent -- the extra signatures could be invalid but verification would pass.  I don't think we want this.

That is, if there are `m` signatures required, and there are `n - m` public keys given, but `m + k` signatures provided, then exactly `k` distinct signatures must recover to `k` distinct public keys out of the `n - m` public keys given.  Also, what happens if `m + k > n` -- i.e. you recover more than the `n` public keys that went into the address hash bytes?

jcnelson replied:
If it simplifies things, we could simply require _exactly_ `m` signatures like we do in the other multisig spending conditions.  That's at least consistent with the multisig experience in other blockchains.

jbencin replied:
I've added some tests for this (It's a PR to Vlad's branch, but will be merged soon), and it works as expected:
 
| Condition                                | Valid? |
|------------------------------------------|--------|
| Extra sigs from allowed signers          | ✓      |
| Extra sigs from random signers           | ✗      |
| Number of signatures below threshold     | ✗      |

fess-v replied:
merged @jbencin @jcnelson

jcnelson commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -567,38 +568,86 @@ impl StacksBlock {
     pub fn validate_transactions_static_epoch(
         txs: &[StacksTransaction],
         epoch_id: StacksEpochId,
+        quiet: bool,
```
Can you take the `quiet` out altogether?  Errors should always be logged.

jcnelson commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -567,38 +568,86 @@ impl StacksBlock {
                 }
             }
+            match &tx.auth {
+                TransactionAuth::Sponsored(ref origin, ref sponsor) => {
```
There might be an opportunity to DRY this up a bit.  You could implement a "supported in epoch?" check for `TransactionSpendingCondition` itself, and then call that here.

fess-v replied:
tx.auth.is_supported_in_epoch @jcnelson added to TransactionAuth instead of TransactionSpendingCondition to move the whole condition there, does it look good now?

jcnelson commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -804,7 +853,8 @@ impl StacksMessageCodec for StacksMicroblock {
         let txs: Vec<StacksTransaction> = {
             let mut bound_read = BoundReader::from_reader(fd, MAX_MESSAGE_LEN as u64);
-            read_next(&mut bound_read)
+            // The latest epoch where StacksMicroblock exist is Epoch25
```
What do microblocks have to do with the code here?  Also, you can use `read_next()` since you're using a `BoundReader`

jbencin commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -314,6 +315,204 @@ impl MultisigSpendingCondition {
+
+    fn consensus_deserialize<R: Read>(
+        fd: &mut R,
+    ) -> Result<OrderIndependentMultisigSpendingCondition, codec_error> {
```
See this discussion from September ([link](https://github.com/stacks-network/stacks-core/pull/3710#discussion_r1330289607)). I made the same suggestion and @fess-v implemented it, but we later reverted it because you didn't want to risk making a change to consensus-critical code

jcnelson replied:
Ah, you're right.  My past self once again saves my current self from being reckless

jcnelson commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -1387,6 +1388,35 @@ impl TransactionAuth {
         }
     }
+
+    pub fn is_supported_in_epoch(
```
Thanks for this.

1. Can you add a docstring?
2. Can you add a dedicated unit test for it?

fess-v replied:
added @jcnelson

jcnelson commented on stackslib/src/chainstate/stacks/auth.rs:
```diff
@@ -557,14 +763,60 @@ impl TransactionSpendingCondition {
+
+        Some(TransactionSpendingCondition::OrderIndependentMultisig(
+            OrderIndependentMultisigSpendingCondition {
+                signer: signer_addr.bytes.clone(),
```
Is this `.clone()` needed?

jbencin replied:
Good catch! There are a number of similar functions that I will also fix

jcnelson commented on stackslib/src/chainstate/stacks/block.rs:
```diff
@@ -599,6 +600,10 @@ impl StacksBlock {
                 }
             }
+            if !tx.auth.is_supported_in_epoch(epoch_id) {
+                error!("Order independent multisig transactions not supported before Stacks 3.0");
```
Tweak -- this error messasge should be more generic (e.g. "this authentication mode is not supported in {epoc_hid}")

jcnelson commented on stackslib/src/chainstate/stacks/db/transactions.rs:
```diff
@@ -1480,6 +1480,15 @@ impl StacksChainState {
             }
         }
 
+        if !StacksBlock::validate_transactions_static_epoch(&vec![tx.clone()], epoch) {
```
This is a somewhat expensive check, and it's already performed by the time this code is called.  Can you instead update `StacksChainState::process_transaction_precheck()` to (1) take a `StacksEpochId` argument, and (2) verify that the `.auth` field is supported in the epoch?  It should be the very first check -- before the call to `tx.verify()`

jcnelson commented on stackslib/src/chainstate/stacks/db/transactions.rs:
```diff
@@ -9439,6 +9448,28 @@ pub mod test {
         let mut chainstate =
             instantiate_chainstate_with_balances(false, 0x80000000, function_name!(), balances);
 
+        let mut tx_runtime_checkerror_trait_no_version = StacksTransaction::new(
```
I'm glad you added these new tests!  With the requested change to `process_transaction()` to remove the call to `StacksBlock::validate_transactions_static_epoch()`, you can preserve them simply by explicitly calling `StacksBlock::validate_transactions_static_epoch()` on these faulty transactions.

jcnelson commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -1228,12 +1244,109 @@ mod test {
     use crate::net::*;
 
+    impl StacksTransaction {
+        fn sign_no_append_origin(
```
Can you add some rustdocs to these functions?

jcnelson reviewed (commented):
This PR is looking pretty good :tada:  Mostly minor comments.  I'll approve once they're addressed.

fess-v commented:
@jbencin @jcnelson merged develop, we added fixes for all latest comments, only 3 unresolved conversations left which wait for @jcnelson reply

jcnelson commented on stackslib/src/chainstate/stacks/miner.rs:
```diff
@@ -2592,6 +2594,19 @@ impl BlockBuilder for StacksBlockBuilder {
                 return TransactionResult::problematic(&tx, Error::NetError(e));
             }
+
+            if !StacksBlock::validate_transactions_static_epoch(&[tx.clone()], epoch_id) {
```
Remove this from this code-path please.

jcnelson reviewed (dismissed):
Well done seeing this through to the end @fess-v!

Let's get that rider SIP added to SIP-021 so this activates in the next hard fork :)

jcnelson commented:
Before merging, I would like @kantai or @obycode to take a look as well.

obycode commented on stackslib/src/chainstate/stacks/transaction.rs:
```diff
@@ -1228,12 +1244,117 @@ mod test {
     use crate::chainstate::stacks::{
         StacksPublicKey as PubKey, C32_ADDRESS_VERSION_MAINNET_MULTISIG,
-        C32_ADDRESS_VERSION_MAINNET_SINGLESIG, *,
+        C32_ADDRESS_VERSION_MAINNET_SINGLESIG, *, *,
```
```suggestion
        C32_ADDRESS_VERSION_MAINNET_SINGLESIG, *,
```

obycode reviewed (dismissed):
This LGTM. Just two minor comments.

fess-v commented on stackslib/src/chainstate/stacks/db/blocks.rs:
```diff
@@ -6663,7 +6665,14 @@ impl StacksChainState {
 
-        // 4: the account nonces must be correct
+        // 4: check if transaction is valid in the current epoch
+        if !StacksBlock::validate_transactions_static_epoch(&[tx.clone()], epoch) {
```
@obycode 
![image](https://github.com/stacks-network/stacks-core/assets/26626714/8f1b786d-ba26-4cb3-a3cd-4e147dd0948f)
it was like that before the PR too
IMO we need to either refactor all these functions on a screenshot or leave them consistent with each other

obycode replied:
This clone is new, right? You could leave the outer function as-is, matching the others, but modify its implementation to call a new function that just validates one transaction.

The call here would be:
```rust
        // 4: check if transaction is valid in the current epoch
        if !StacksBlock::validate_transaction_static_epoch(tx, epoch) {
```

And then we would have:
```rust
    /// Verify that all transactions are supported in the given epoch, as indicated by `epoch_id`
    pub fn validate_transactions_static_epoch(
        txs: &[StacksTransaction],
        epoch_id: StacksEpochId,
    ) -> bool {
        txs.iter().all(|tx| validate_transaction_static_epoch(tx, epoch_id))
    }
```

obycode reviewed (approved):
Thanks!

will-corcoran commented:
Amazing work.  To echo @jcnelson  - kudos on your perseverance @fess-v !

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
